### PR TITLE
[DLSP21] Week 12-2

### PR DESCRIPTION
**Pull request by angelaaaateng**
Adding class notes from Week 12 part 2,  guest lecture

Written by Angela Teng (part 1) and Joanna Jin (part 2); hackmd.io can be found here https://hackmd.io/@q1HeL2WPSUmFYugZ4SM7Xg/12_2

Please let us know if we should change or revise our notes, and we will be happy to incorporate your feedback. Thank you!

### CODE REVIEW

**Comment by Atcold** [comment]
There are no figures.

**Comment by angelaaaateng** [comment]
Hi Professor @Atcold ! We added the figures in based on the documentation you sent, please let us know if this looks good to you or if you have additional edits :)

**Review by Atcold — CHANGES_REQUESTED** [review]
Plz, fix. Everything's broken now.

```
   Server address: http://127.0.0.1:4000/NYU-DLSP21/
  Server running... press ctrl-c to stop.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure4.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure1.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure5.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure3.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure2.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure6.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure7.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure9.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure11.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure12.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure10.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure8.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure13.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure14.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure17.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure16.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure15.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure18.png' not found.
[2021-05-19 01:20:07] ERROR `/NYU-DLSP21/en/week12/12-2/{​{site.baseurl}}/images/week12/12-2/figure19.png' not found.
```

File: docs/en/week12/12-2.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+title: Low Resource Machine Translation
+lecturer: Marc’Aurelio Ranzato
+authors: Angela Teng, Joanna Jin
+date: 17 May 2021
+---
+
+# 12.2
+## Wed 21 Apr – (Invited) Lecture 12: Guest Lecture by Marc'Aurelio Ranzato
+
+This week's lecture was a guest lecture by [Marc'Aurelio Ranzato](https://ai.facebook.com/people/marc-aurelio-ranzato/), who is a research scientist and manager at the Facebook AI Research (FAIR) lab, where he works to enable machines to learn with weaker supervision and to efficiently transfer knowledge across tasks. The first part of this document focuses on understanding low resource machine translation, and the second half discusses potential domain mismatches in machine learning and machine translation.
```
Comment (Atcold) [inline docs/en/week12/12-2.md:12]:
```suggestion
```

File: docs/en/week12/12-2.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+## ML Perspective on Low Resource MT
```
Comment (Atcold) [inline docs/en/week12/12-2.md:14]:
```suggestion
```

File: docs/en/week12/12-2.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+We start with standard machine learning algorithms can be applied in the realm of machine translation translation. Then, we take a deeper look into understanding different perspectives on this application. Our setting consists of multiple languages and multiple domains, but eventually we want to maximize translation accuracy of a certain domain in a language class.
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-2/figure1.png" style="background-color:#DCDCDC;" /><br>
```
Comment (Atcold) [inline docs/en/week12/12-2.md:18]: Where's the caption? No figure has a caption.

File: docs/en/week12/12-2.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+## Case Studies
+**Translating from English to Nepali:** Let's start with a simple case study within the realm of supervsied learning. For example, let's say that we have a sentence in English and we want to trasnslate it to Nepali. This is similar to multitask learning in the sense that you have one task, and then you add another task you're interested in. Since we have multiple domains we can start thinking about domain adaptation techniques and analyzing which domain adaptation techniques are applicable to machine translation.
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-2/figure2.png" style="background-color:#DCDCDC;" /><br>
+</center>
+
+Beginning with a supervised learning setting, we can use the aforementioned example of translation. Our dataset is
+$$D = \{(x,y)_i\}_{i=1, ..., N}$$
```
Comment (Atcold) [inline docs/en/week12/12-2.md:32]: You need an empty line before and after an equation.
Check out [the instructions](https://atcold.github.io/pytorch-Deep-Learning/en/about/).

Comment (Atcold) [reply docs/en/week12/12-2.md:32]: Moreover, use `\vect{}` for all vectors in the text, here and elsewhere.

File: docs/en/week12/12-2.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+We can train this model using maximum likelihood, where we want to maximize $\{ y\vert x \} $.
```
Comment (Atcold) [inline docs/en/week12/12-2.md:34]: The conditional bar is `\mid`.
Curly brackets are `\lbrace` and `\rbrace`.

File: docs/en/week12/12-2.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+One way to represent this methodology is by the diagram, where you have a blue encoder that processes English sentences, and a red decoder that processes Nepali. From here, we want to compute our cross-entropy loss and update our model parameters. We may also want to regularize the model using either dropout or label smoothing. Additionally, we may need to regularize using the log loss
+
+$$-lo gp(y|x; \theta )$$
```
Comment (Atcold) [inline docs/en/week12/12-2.md:38]: You forgot a `\` before `log`.
Vectors must be enclosed in `\vect{}`. Here and everywhere.

File: docs/en/week12/12-2.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+<img src="{{site.baseurl}}/images/week12/12-2/figure3.png" style="background-color:#DCDCDC;" /><br>
+</center>
+
+One way to leverage an additional dataset is by trying to model $p(x)$ with a semi-supervised approach. One way to model $$p(x)$$ is via the denoising of an auto-encoder. This is particualrly useful because the encoder and decoder share a simlar machine translation methodology.
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-2/figure4.png" style="background-color:#DCDCDC;" /><br>
+</center>
+
+Another approach would be via self-training or pseudolearning. This is an algorithm from the 90s and the idea is that you take your sentence from the monolingual dataset and then you inject noise to it, you encode the poor translation you made and then make a prediction to produce the desired outcome. You can tune the parameters by minimizing the standard cross entropy loss on the labeled data. In other words, we are minimizing $L^{sup} (\theta) + \lambda L^{ST} \theta$. We are basically using a stale version of our model to produce the desired output $y$.
```
Comment (Atcold) [inline docs/en/week12/12-2.md:54]: You forgot a `\` before `sup`.

Comment (Atcold) [reply docs/en/week12/12-2.md:54]: I think you want to enclose `ST` in `\text{}`.

File: docs/en/week12/12-2.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-2/figure5.png" style="background-color:#DCDCDC;" /><br>
```
Comment (Atcold) [inline docs/en/week12/12-2.md:57]: Write the math and text. This is not a figure.

File: docs/en/week12/12-2.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-2/figure6.png" style="background-color:#DCDCDC;" /><br>
+</center>
+
+The algorithm would stay the same as above, where we have smaller systems of encoders and decoders.
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-2/figure7.png" style="background-color:#DCDCDC;" /><br>
+</center>
```
Comment (Atcold) [inline docs/en/week12/12-2.md:74]: Write the math and text. These are not a figure.

File: docs/en/week12/12-2.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+**How can we improve generalization?**
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-2/figure3.png" style="background-color:#DCDCDC;" /><br>
+</center>
+
+One way to leverage an additional dataset is by trying to model $p(x)$ with a semi-supervised approach. One way to model $$p(x)$$ is via the denoising of an auto-encoder. This is particualrly useful because the encoder and decoder share a simlar machine translation methodology.
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-2/figure4.png" style="background-color:#DCDCDC;" /><br>
```
Comment (Atcold) [inline docs/en/week12/12-2.md:51]: Remove the math from the screenshot and type it out.
Figures are meant to be used for diagrams.
We have LaTeX for the math.
Here and elsewhere.

File: docs/en/week12/12-2.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+* STDM: a new kind of domain mismatch, intrinsic to the MT task.
+* STDM is particularly significant in low resource language pairs.
+* Metric & controlled setting enable study and better understanding of STDM.
+* Methods that leverage source side monolingual data are more robust to STDM.
+* In practice, the influence of STDM depends on several factors, such as the amount of
+parallel and monolingual data, the domains, language pair, etc. In particular, if domains
+are not too distinct, STDM may even help regularizing!
+
+
+Prepared by Angela Teng (part 1) and Joanna Jin (part 2)
```
Comment (Atcold) [inline docs/en/week12/12-2.md:184]:
```suggestion
```

**Review by Atcold — CHANGES_REQUESTED** [review]
You haven't added the summary to `en/week12/12.md`.
Do not use figures for text or math. Type the math in LaTeX.
No figure has a caption.
Use `\vect{}` for **all** vectors in the write up.
The conditional bar is `\mid`.

**Comment by angelaaaateng** [comment]
Hi Professor @Atcold ! Just incorporated your feedback, thanks so much for taking the time :) Please let us know what your thoughts are on our latest commit and if you have further revisions. Thank you!

File: docs/en/week12/12-2.md [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+
+
+## Case Studies
+**Translating from English to Nepali:** Let's start with a simple case study within the realm of supervsied learning. For example, let's say that we have a sentence in English and we want to trasnslate it to Nepali. This is similar to multitask learning in the sense that you have one task, and then you add another task you're interested in. Since we have multiple domains we can start thinking about domain adaptation techniques and analyzing which domain adaptation techniques are applicable to machine translation.
+
+We can define our supervised learning method as the following when translating from English to Nepali:
+
+$$\mathcal{D}=\left\{\left(x_{\mathbf{1}} y\right)_{i}\right\}_{i=1, \ldots, N}$$
```
Comment (Atcold) [inline docs/en/week12/12-2.md:27]: Why `\mathbf{1}`?
Also, please, all `x`, `y`, `θ` should be encapsulated by `\vect{}`.

File: docs/en/week12/12-2.md [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+<center>
+<img src="{{site.baseurl}}/images/week12/12-2/figure2_2.png" style="background-color:#DCDCDC;" /><br>
+<b>Figure 2:</b> Supervised Learning Visualization
+</center>
+
+We can train this model using maximum likelihood, where we want to maximize $\{ y\vert x \} $.
+
+One way to represent this methodology is by the diagram, where you have a blue encoder that processes English sentences, and a red decoder that processes Nepali. From here, we want to compute our cross-entropy loss and update our model parameters. We may also want to regularize the model using either dropout or label smoothing. Additionally, we may need to regularize using the log loss
+
+$$-\log p(y|x; \theta )$$
```
Comment (Atcold) [inline docs/en/week12/12-2.md:55]: The conditional bar is `\mid`, here and everywhere in this document.

File: docs/en/week12/12-2.md [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+
+$$\mathcal{D} =\lbrace {(x, y)_{i} \rbrace }_{i=1, \vect{...}, N}$$
+
+$$\mathcal{M}^{s}=\left\{x_{j}^{s}\right\}_{j=1, \vect{...}, M_{s}}$$
+
+And when we want to predict missing labels using decoding and training noise, we can use cross-entropy loss as such:
+
+$\mathcal{L}^{S T}(\theta)=-\log p(\bar{y} \mid x+n)$
+
+$\mathcal{L}(\theta)=\mathcal{L}^{\sup }(\theta)+\lambda \mathcal{L}^\text{S T}(\theta)$
```
Comment (Atcold) [inline docs/en/week12/12-2.md:96]: Maybe you wanted to use a doulble `$`?
Also, use `\text{ST}` both times, and without a space between S and T.

File: docs/en/week12/12-2.md [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+<center>
+<img src="{{site.baseurl}}/images/week12/12-2/figure9.png" style="background-color:#DCDCDC;" /><br>
+<b> Figure 6:</b> Semi-Supervised Learning Case Scenario on Multiple Languages
+</center>
+
+This is the learning framework for multilingul training, which shares encoder and decoder across all the language pairs, prepends a target language identifier to the source sentence to inform decoder of desired language and concatenates all the datasets together.
+
+Mathematically, the train uses standard cross-entropy loss:
+
+$$\mathcal{L}(\theta)=-\sum_{s,t} \mathbb{E}_{(x,y)\sim\mathcal{D}}[log p(y|x,t)]$$
```
Comment (Atcold) [inline docs/en/week12/12-2.md:176]: The log needs a `\` before. Here and everywhere else. Do use find-and-replace.

File: docs/en/week12/12-2.md [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+date: 17 May 2021
+---
+
+# 12.2
+## Wed 21 Apr – (Invited) Lecture 12: Guest Lecture by Marc'Aurelio Ranzato
+
+We start with standard machine learning algorithms can be applied in the realm of machine translation translation. Then, we take a deeper look into understanding different perspectives on this application. Our setting consists of multiple languages and multiple domains, but eventually we want to maximize translation accuracy of a certain domain in a language class.
+
+<center>
+<img src="{{site.baseurl}}/images/week12/12-2/figure1.png" style="background-color:#DCDCDC;" /><br>
```
Comment (Atcold) [inline docs/en/week12/12-2.md:15]: Plz, replace with text.

**Review by Atcold — CHANGES_REQUESTED** [review]
Almost there, a few more edits and you're done.

**Comment by angelaaaateng** [comment]
Hi Professor @Atcold ! Thanks so much for the detailed feedback here, please see the latest changes. Thanks so much!

File: docs/en/week12/12-2.md [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+
+
+## Case Studies
+**Translating from English to Nepali:** Let's start with a simple case study within the realm of supervsied learning. For example, let's say that we have a sentence in English and we want to trasnslate it to Nepali. This is similar to multitask learning in the sense that you have one task, and then you add another task you're interested in. Since we have multiple domains we can start thinking about domain adaptation techniques and analyzing which domain adaptation techniques are applicable to machine translation.
+
+We can define our supervised learning method as the following when translating from English to Nepali:
+
+$$\mathcal{D}=\left\{\left(\vect{x}_{1} \vect{y}\right)_{i}\right\}_{i=1, \ldots, N}$$
```
Comment (Atcold) [inline docs/en/week12/12-2.md:35]: I still don't get it. What is `x1y`???

File: docs/en/week12/12-2.md [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+
+Our per-sample loss using the usual attention-based transformer is defined as the following:
+
+$$\mathcal{L}(\vect{\theta})=-\log p(\vect{y \mid \vect{x} ; \vect{\theta})$$
```
Comment (Atcold) [inline docs/en/week12/12-2.md:39]: image

File: docs/en/week12/12-2.md [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+\begin{array}{c}
+\mathcal{D}=\left\{(\vect{x}, \vect{y})_{i}\right\}_{i=1, \ldots, N} \\
+\mathcal{M}^{s}=\left\{\vect{x}_{j}^{s}\right\}_{j=1, \ldots, M_{s}}
+\end{array}
+$$
+
+When using the DAE learning framework, we an either pre-train or add a DAE loss to the supervised cross entropy term:
+
+$$
+\mathcal{L}^{D A E}(\vect{\theta})=-\log p(\vect{x} \mid \vect{x}+n)
```
Comment (Atcold) [inline docs/en/week12/12-2.md:86]: Use `\text{}` for DAE. Also, no spaces.

File: docs/en/week12/12-2.md [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+<b>Figure 5:</b> Semi-Supervised Learning (ST)
+</center>
+
+Another approach would be via self-training or pseudolearning. This is an algorithm from the 90s and the idea is that you take your sentence from the monolingual dataset and then you inject noise to it, you encode the poor translation you made and then make a prediction to produce the desired outcome. You can tune the parameters by minimizing the standard cross entropy loss on the labeled data. In other words, we are minimizing $L^{\sup} (\vect{\theta}) + \lambda L^\text{ST} \vect{\theta}$. We are basically using a stale version of our model to produce the desired output $\vect{y}$.
+
+**ALGORITHM**
+
+- train model $p(\vect{y} \mid \vect{x})$ on $D$
+- repeat
+  - decode $\vect{x}^s \sim \mathcal{M}^s$ to $\bar{\vect{y}}$ and create additional dataset $\mathcal{A}^s = \{(\vect{x}^s_j,\vect{y}_j\}_{j=1,\vect{...},M_s})$
```
Comment (Atcold) [inline docs/en/week12/12-2.md:117]: image

File: docs/en/week12/12-2.md [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+$$
+
+We can get additional source and target side monolingual data.
+
+The process we can follow is via this algorithm:
+- train model $p(\vect{x} \mid \vect{y})$ and $p(\vect{y} \mid \vect{x})$ on $\mathcal{D}$
+- repeat
+  - *Phase 1*
+  - decode $\vect{y}^{t} \sim \mathcal{M}^{t}$ to $\bar{\vect{x}}$ with $p(\vect{x} \mid \vect{y})$, create additional dataset $$\mathcal{A}^{t}=\left\{\left(\bar{\vect{x}}_{k}, \vect{y}_{k}^{t}\right)\right\}_{k=1, \ldots, M_{t}}$$
+  - decode $\vect{x}^{s} \sim \mathcal{M}^{s}$ to $\bar{\vect{y}}$ with $p(\vect{y} \mid \vect{x})$, create additional dataset $$\mathcal{A}^{s}=\left\{\left(\vect{x_{j}^{s}, \bar{\vect{y}}_{j}\right)\right\}_{j=1, \ldots, M}$$
```
Comment (Atcold) [inline docs/en/week12/12-2.md:160]: image

**Review by Atcold — CHANGES_REQUESTED** [review]
Please, make sure the LaTeX displays correctly.
After this we should be fine.

File: docs/en/week12/12-2.md (hunk 17 above)
Comment (angelaaaateng) [reply docs/en/week12/12-2.md:35]: Apologies, Professor that was supposed to be a comma, my bad on this 😅  Thanks for pointing it out, I have revised it and will commit along with the LaTex edits.

**Comment by angelaaaateng** [comment]
Hi Professor @Atcold  Thanks very much for your patience and detailed feedback, really appreciate it! :) please see the latest revisions in the most recent commit. Thanks for your time on this.

**Review by Atcold — APPROVED** [review]
Very good, thanks! 💜

**Pull request merged**